---
title: "R Notebook"
output:
  html_document:
    df_print: paged
---
```{r, setup}
#for 3d pca
#knitr::knit_hooks$set(webgl = hook_webgl)
```

 Generate data frames 
 RBP Feature plots 
 RMATS pipeline 
 Combined RMATS events 
 Sashimi with feature 
 RMAPS combined 
 RMAPS with feature 
 Ven diagram 

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~CREATE DATA    V
```{r fig.height=15, fig.width=15, warning = FALSE, message = FALSE}
source("ccRCC_Wang_2020_funcrtions.R")

#create counts matrix using STAR results
if (file.exists("Table counts.csv")){ctstk <- read.csv(file = "Table counts.csv", row.names = 1)
}else{ print('Counts file not found')}

#create meta data table
meta <- meta_table_gen(meta_path = "Data from Wang et al/Table meta_build1.csv",
                       meta_more_path = "Data from Wang et al/Table meta_build2.csv")

rownames(meta) <- meta$Sample.Name

#Change to Sample name instead of Run
if (identical(meta[order(meta[,"Run"]), "Run"], colnames(ctstk)))
{colnames(ctstk) <- meta[order(meta[, "Run"]), "Sample.Name"]}

#sub-tables
cts_c_t <- ctstk[,meta$State != "Normal"]
cts_pt_tt <- ctstk[,meta$Thrombosis == "Yes"]

if(!exists("wrs_res"))
{ wrs_res <- wrs_res_gen(ctstk, meta)} #Wilcoxon rank sum test for differential expressed genes

annotation_color_list <- ann_colors() #for future heatmaps

gene_lists <- gene_lists_gen(ctstk, wrs_res) #for pca and more

state_levels <- c('Normal','Cancer','Primary Tumor','Thrombus Tumor') #annotation

Ann_all <- meta[,c('sex','Age','T.stage','N.stage','M.stage','MayoLevel','State','Thrombosis', 'Provider')]
  rownames(Ann_all) <- meta$Sample.Name
  
if (!exists("dsq_normal_cancer"))
{
  dsq_normal_cancer <- DESeqDataSetFromMatrix(countData = ctstk,
                                              colData = meta,
                                              design= ~ Normal)
  dsq_normal_cancer <- DESeq(dsq_normal_cancer)
  cts_normal_cancer <- counts(dsq_normal_cancer,normalized=TRUE)
  res_normal_cancer <- results(dsq_normal_cancer,contrast = c('Normal','Yes','No'))
  if(!file.exists("Table counts normalized.csv")){
    write.csv(cts_normal_cancer, file = "Table counts normalized.csv")}
  
  dsq_nc_t <- DESeqDataSetFromMatrix(countData = ctstk,
                                              colData = meta,
                                              design= ~ Thrombosis)
  dsq_nc_t <- DESeq(dsq_nc_t)
  cts_nc_t <- counts(dsq_nc_t,normalized=TRUE)
  res_nc_t <- results(dsq_nc_t,contrast = c('Thrombosis','Yes','No'))
}
  
#######RMATS ORDER ALL#####
  x <- read.csv('Table rmats_order.csv', row.names = 1)
  rmats_order_all <- x$x
  names(rmats_order_all) <- rownames(x)

######DECONV META##########
ref.based.estimates <- read.csv(file = "Table deconvolution.csv", row.names = 1)
cnv_hm <- ref.based.estimates # For heatmap
cnv_hm_breaks <- quantile_breaks(cnv_hm,101) # for better coloring
cnv_hm[is.na(cnv_hm)] <- 0 # na = 0

#annotations above heatmap for better order,  rownames(ref.based.estimates) are cell types
ann_all_col <- c('sex','Age', 'MayoLevel','T.stage', 'N.stage','M.stage',
                 rownames(ref.based.estimates), 'State', 'Thrombosis', 'Provider')

#df for annotation colors
Ann_all_t <- Ann_all #Ann_all is a df for annotation. rows are sample, columns are annotation to add
annotation_color_list_t <- ann_colors() #color of other annotations

#color range 0 to 1 , can be also according to qunatile_breaks
cvm_col <- colorRamp2(seq(0,1,length.out = nrow(cnv_hm)),#cnv_hm_breaks, 
                       colorRampPalette(c('ghostwhite', '#FFFFBF','firebrick'))(nrow(cnv_hm)))

#add color range for each cell type, in range of min and max for that cell type
for (row in rownames(ref.based.estimates))
{
  Ann_all_t[,row] <- as.numeric(cnv_hm[row,]) #change color by seq or [0,1]
  annotation_color_list_t[[row]] <- colorRamp2(seq(min(ref.based.estimates[row,]),
                                                   max(ref.based.estimates[row,]),
                                                   length.out = nrow(cnv_hm)),#cnv_hm_breaks, 
                       colorRampPalette(c('ghostwhite', '#FFFFBF','firebrick'))(nrow(cnv_hm)))#cvm_col
}

#Ann_all_t[,'Provider'] <- meta$BIOMATERIAL_PROVIDER

rmats_events_combined_inc <- read.csv('Table S1 inclusion levels.csv', row.names = 1)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~HEATMAP WITH CELL TYPE   V
```{r}
for_complexHM <- as.matrix(cts_normal_cancer[unique(gene_lists$Splicosome),])
for_complexHM <- for_complexHM[apply(for_complexHM, 1, var, na.rm=TRUE) != 0,] # remove genes with no variance

cell_h <- 1
cell_w <- 1
plot_name <- 'Spliceosome'

png(paste0('plots/',plot_name,'.png'),
  width = cell_w*ncol(for_complexHM)+150,
  height = cell_h*nrow(for_complexHM)+250,
  units = "mm",
  res = 300)
complex_hm_breaks <- quantile_breaks(scale_rows(for_complexHM),100)
cHM <- Heatmap(t(scale(t(for_complexHM))),
        column_title = plot_name,
        name = 'Expression level',
        col = colorRamp2(complex_hm_breaks, colorRampPalette(c("skyblue","#FFFFBF", "#D73027"))(length(complex_hm_breaks))),
        na_col = 'black',
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        clustering_method_rows = "ward.D",
        clustering_method_columns = "ward.D",
        width = ncol(for_complexHM) * unit(cell_w,'mm'),
        height = nrow(for_complexHM) * unit(cell_h,'mm'),
        row_names_gp = gpar(fontsize = 4),
        column_names_gp = gpar(fontsize = 3),
        row_names_max_width = max_text_width(rownames(for_complexHM)),
        heatmap_legend_param = list(direction = 'horizontal', legend_width  = unit(2,'inches')),
        top_annotation = HeatmapAnnotation( df = Ann_all_t[,ann_all_col],
                                            col = annotation_color_list_t,
                                            na_col = 'white',
                                            annotation_name_gp= gpar(fontsize = 12),
                                            simple_anno_size = unit(4, "mm"),
                                            show_legend = c(rep(TRUE,6),rep(FALSE,29),TRUE,TRUE,TRUE),
                                            gp = gpar(col = "grey90"),
                                            # To have 1 colorbar for all celltypes instead of 1 for each 
                                            annotation_legend_param =
                                            list('State' = list(at = c('Normal', 'Cancer', 'Thrombus')),
                                                 'Type B intercalated cell' = list(title = 'Cell inclusion',
                                                                                   labels = c("Min", "Half", "Max"))))
        
    )

draw (cHM,heatmap_legend_side = "bottom",annotation_legend_side = "left", legend_grouping = "original")
dev.off()
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PCA SIMPLE   V
```{r}
pca_simple <- pca_plot(cts_normal_cancer,
                     groupMeta = meta,
                     meta_type = 0,
                     pc1 =2,
                     pc2 = 1,
                     myTitle = 'PCA'
                     ) +
  guides(fill = guide_legend(label.position = "right"), keywidth=1) + 
  theme_bw() + 
  theme(plot.title = element_text(size=40, hjust = 0.5),
        legend.box.background = element_rect(colour = "black"),
        legend.position = 'bottom',
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))

ggsave("plots/PCA_Simple.png", width = 15, height = 15, limitsize = FALSE)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PCA FEATURE PLOT (COUNTS/RMATS)   V
```{r}
#df with events
rmats_pca <- (na.omit(rmats_events_combined_inc[,colnames(cts_normal_cancer)]))

#event id
gene<- c('RMATS')
event_ix <- 100
gene_feature_plot <- feature_plot2(rmats_pca,
                                  inclusion_level = as.numeric(rmats_pca[event_ix,]),
                                  myTitle = paste0('rmats combined ','.'))

ggsave(paste0('plots/FEATURE_',gene,'.png'), 
         plot = gene_feature_plot, width = 15, height = 15)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PCA RMATS    V
```{r}
pca_simple<-pcaMethods3(dsq_normal_cancer,
           df = na.omit(rmats_events_combined_inc[,colnames(cts_normal_cancer)]),
           intgroup = c("State", 'Thrombosis'),
           pc1=1, pc2=2 ,
           pCutoff = 0.01,
           FCcutoff = 1,
           pSize=2,
           lSize = 1,
           labStyle = 2, # 1no / 2text / 3box / 4box + arrow

           toScale=T,
           lColor = 0,
           #marking1 = c('C011c','C011t','C012c','C012t','C013c','C013t','C014c','C014t','C015c','C015t','C016c','C016t'),
           #marking2 = c('C030c','C030t','C031c','C031t','C032c','C032t','C033c','C033t','C034c','C034t','C035c','C035t','C036c','C036t'),
           incLevel_size = 1,
           myTitle = 'PCA (RMATS events combined)',
           groupMeta = meta) + theme_bw()

ggsave("plots/PCA_RMATS1.png", width = 10, height = 10)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PCA WITH ARROW, DUMB WAY   V
```{r fig.height=15, fig.width=15}

df_arw <- cts_normal_cancer[, meta$Thrombosis == 'Yes']
df_arw <- df_arw[!apply(df_arw,1,function(x) min(x) == max(x)),]
df_pca_arw <- prcomp(x = t(df_arw),
               scale. = T,
               center = T)

percentVar <- df_pca_arw$sdev^2 / sum( df_pca_arw$sdev^2 )
xs = paste0("PC1: ", round(percentVar[1] * 100),"% variance")
ys = paste0("PC2: ", round(percentVar[2] * 100),"% variance")

df_pca_arw <- data.frame( x = df_pca_arw$x[,"PC1"],y = df_pca_arw$x[,"PC2"],state = patient_state(meta[meta$Thrombosis == 'Yes',]))
df_pca_arw <- df_pca_arw[order(rownames(df_pca_arw))[1:60],]
row_odd <- seq_len(60) %% 2



df_arw_part <- geom_segment (aes(x = df_pca_arw[row_odd == 0,'x'],
                                 y = df_pca_arw[row_odd == 0,'y'],
                                 xend = df_pca_arw[row_odd == 1,'x'],
                                 yend = df_pca_arw[row_odd == 1,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65)

df_pnt_part <- ggplot() + 
  geom_point(data = df_pca_arw,aes(x = x, y = y,  color = state),size = 7, alpha = 1.0)   +
  geom_text(data = df_pca_arw, aes(x = x,   y = y), label = rownames(df_pca_arw), color = "black",size = 3)

pca_simple_arrow <- df_pnt_part + #pca_simple + 
  geom_segment (aes(x = df_pca_arw[1,'x'],y = df_pca_arw[1,'y'],xend = df_pca_arw[2,'x'],yend = df_pca_arw[2,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[3,'x'],y = df_pca_arw[3,'y'],xend = df_pca_arw[4,'x'],yend = df_pca_arw[4,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[5,'x'],y = df_pca_arw[5,'y'],xend = df_pca_arw[6,'x'],yend = df_pca_arw[6,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[7,'x'],y = df_pca_arw[7,'y'],xend = df_pca_arw[8,'x'],yend = df_pca_arw[8,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[9,'x'],y = df_pca_arw[9,'y'],xend = df_pca_arw[10,'x'],yend = df_pca_arw[10,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[11,'x'],y = df_pca_arw[11,'y'],xend = df_pca_arw[12,'x'],yend = df_pca_arw[12,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[13,'x'],y = df_pca_arw[13,'y'],xend = df_pca_arw[14,'x'],yend = df_pca_arw[14,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[15,'x'],y = df_pca_arw[15,'y'],xend = df_pca_arw[16,'x'],yend = df_pca_arw[16,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[17,'x'],y = df_pca_arw[17,'y'],xend = df_pca_arw[18,'x'],yend = df_pca_arw[18,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[19,'x'],y = df_pca_arw[19,'y'],xend = df_pca_arw[20,'x'],yend = df_pca_arw[20,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[21,'x'],y = df_pca_arw[21,'y'],xend = df_pca_arw[22,'x'],yend = df_pca_arw[22,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[23,'x'],y = df_pca_arw[23,'y'],xend = df_pca_arw[24,'x'],yend = df_pca_arw[24,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[25,'x'],y = df_pca_arw[25,'y'],xend = df_pca_arw[26,'x'],yend = df_pca_arw[26,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[27,'x'],y = df_pca_arw[27,'y'],xend = df_pca_arw[28,'x'],yend = df_pca_arw[28,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[29,'x'],y = df_pca_arw[29,'y'],xend = df_pca_arw[30,'x'],yend = df_pca_arw[30,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[31,'x'],y = df_pca_arw[31,'y'],xend = df_pca_arw[32,'x'],yend = df_pca_arw[32,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[33,'x'],y = df_pca_arw[33,'y'],xend = df_pca_arw[34,'x'],yend = df_pca_arw[34,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[35,'x'],y = df_pca_arw[35,'y'],xend = df_pca_arw[36,'x'],yend = df_pca_arw[36,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[37,'x'],y = df_pca_arw[37,'y'],xend = df_pca_arw[38,'x'],yend = df_pca_arw[38,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[39,'x'],y = df_pca_arw[39,'y'],xend = df_pca_arw[40,'x'],yend = df_pca_arw[40,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[41,'x'],y = df_pca_arw[41,'y'],xend = df_pca_arw[42,'x'],yend = df_pca_arw[42,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[43,'x'],y = df_pca_arw[43,'y'],xend = df_pca_arw[44,'x'],yend = df_pca_arw[44,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[45,'x'],y = df_pca_arw[45,'y'],xend = df_pca_arw[46,'x'],yend = df_pca_arw[46,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[47,'x'],y = df_pca_arw[47,'y'],xend = df_pca_arw[48,'x'],yend = df_pca_arw[48,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[49,'x'],y = df_pca_arw[49,'y'],xend = df_pca_arw[50,'x'],yend = df_pca_arw[50,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[51,'x'],y = df_pca_arw[51,'y'],xend = df_pca_arw[52,'x'],yend = df_pca_arw[52,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[53,'x'],y = df_pca_arw[53,'y'],xend = df_pca_arw[54,'x'],yend = df_pca_arw[54,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[55,'x'],y = df_pca_arw[55,'y'],xend = df_pca_arw[56,'x'],yend = df_pca_arw[56,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[57,'x'],y = df_pca_arw[57,'y'],xend = df_pca_arw[58,'x'],yend = df_pca_arw[58,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  geom_segment (aes(x = df_pca_arw[59,'x'],y = df_pca_arw[59,'y'],xend = df_pca_arw[60,'x'],yend = df_pca_arw[60,'y']),
                             arrow = arrow(length=unit(.5, 'cm')),colour = "black",alpha = 0.6,lwd = 0.65) +
  theme_bw() + labs (x = xs, y = ys) + 
  theme(plot.title = element_text(size=40, hjust = 0.5),
        legend.box.background = element_rect(colour = "black"),
        legend.position = 'bottom',
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))

ggsave("plots/PCA_SIMPLE_ARROW.png", width = 15, height = 15)

```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RBP FEATURE PLOT   V
```{r}
rbp_feature_save_path <- paste0('plots/RBP_FEATURE_',Sys.Date())
dir.create(rbp_feature_save_path)

 for (rbp_gene in unique(gene_lists$`Motif genes`))
{
   rbp_feature_plot <- feature_plot(cts_normal_cancer,
                                    min_point_size = 0.2,
                                    inclusion_level = as.numeric(cts_normal_cancer[rbp_gene,]),
                                    color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                    myTitle = rbp_gene,
                                    )  +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=30, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
 
 ggsave(paste0(rbp_feature_save_path,'/',rbp_gene,'.png'),
         plot = rbp_feature_plot, width = 15, height = 15)
}
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RMATS LISTS V
```{r fig.width = 15, fig.height = 15}

rmats_all_lists <- c("rmats_lists/ccRCC_wang_2020_onlyN.txt",
                    "rmats_lists/ccRCC_wang_2020_All_CPT.txt",
                    "C0_SE.MATS.JCEC")

rmats_n_c_lists <- c("rmats_lists/ccRCC_wang_2020_N_10.txt",
                     "rmats_lists/ccRCC_wang_2020_Ca_10.txt",
                     "N_C_SE.MATS.JCEC")

rmats_n_pt_lists <- c("rmats_lists/ccRCC_wang_2020_N_10.txt",
                      "rmats_lists/ccRCC_wang_2020_PtTt_10.txt",
                      "N_PT_SE.MATS.JCEC")

rmats_c_pt_lists <- c("rmats_lists/ccRCC_wang_2020_Ca_10.txt",
                      "rmats_lists/ccRCC_wang_2020_PtTt_10.txt",
                      "C_PT_SE.MATS.JCEC")
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RMATS PIPELINE V
```{r}
rmats_compariosn_list <-  list(rmats_n_c_lists,
                               rmats_n_pt_lists,
                               rmats_c_pt_lists
                               )

print ("Starting RMATS pipeline")
for (comparison in rmats_compariosn_list){
print (paste0('Comparison - ', comparison[3]))
rmats_selected_lists <- comparison
rmats_all_table_name <- rmats_all_lists[3]
rmats_selected_table_name <- rmats_selected_lists[3]

  rmats_tables_path <- c("selected" = rmats_selected_table_name,
                         "all" = rmats_all_table_name)

  rmats_fdr_cutoff <- 0
  rmats_inc_diff_cutoff <- 0.2 
  rmats_col_id_1 <- 3
  rmats_col_id_2 <- 11 # MXE is 13
  rmats_table_path_regex <- "(?<=data/rmats/rmats_matrices/).+(?=/(SE|MXE).MATS.(JCEC|JC).txt)"

  #read data
  rmats_table <- read.csv(rmats_tables_path[1], sep='\t')
  rmats_all <- read.csv(rmats_tables_path[2], sep='\t')

  #get order of samples
  rmats_order_all_1_srr <- unlist(str_extract_all(read_lines(rmats_all_lists[1]),"SRR\\d+"))
  rmats_order_all_2_srr <- unlist(str_extract_all(read_lines(rmats_all_lists[2]),"SRR\\d+"))

  rmats_order_selected_1_srr <- unlist(str_extract_all(read_lines(rmats_selected_lists[1]),"SRR\\d+"))
  rmats_order_selected_2_srr <- unlist(str_extract_all(read_lines(rmats_selected_lists[2]),"SRR\\d+"))

  rmats_order_all_1 <- sapply(rmats_order_all_1_srr, function (x) meta$Sample.Name[which(meta$Run == x)])
  rmats_order_all_2 <- sapply(rmats_order_all_2_srr, function (x) meta$Sample.Name[which(meta$Run == x)])
  rmats_order_all <- c(rmats_order_all_1, rmats_order_all_2)

  #change from SRR... to sample name
  rmats_order_selected_1 <- sapply(rmats_order_selected_1_srr, 
                                   function (x) meta$Sample.Name[which(meta$Run == x)])
  rmats_order_selected_2 <- sapply(rmats_order_selected_2_srr, 
                                   function (x) meta$Sample.Name[which(meta$Run == x)])
  rmats_order_selected <- c(rmats_order_selected_1, rmats_order_selected_2)

  #get events with FDR and INC_LEVEL
  rmats_events_id <- rmats_table[abs(rmats_table$IncLevelDifference) > rmats_inc_diff_cutoff
                                 & rmats_table$FDR <= rmats_fdr_cutoff,]
  rmats_all_id <- rmats_all
  rmats_colnames_id <- colnames(rmats_all)[rmats_col_id_1:rmats_col_id_2]

  #create unique ID
  rmats_all_id$event_id <-    apply(rmats_all_id[, rmats_colnames_id, drop = F],
                                    MARGIN = 1,
                                    FUN = function(i) gsub(" ","",paste(i, collapse = "_"),fixed=TRUE))
  rmats_events_id$event_id <- apply(rmats_events_id[, rmats_colnames_id, drop = F],
                                    MARGIN = 1,
                                    FUN = function(i) gsub(" ","",paste(i, collapse = "_"),fixed=TRUE))

  #Find rest of INC_LEVEL by comparing unique ID
  rmats_events_inc <- rmats_all_id[rmats_all_id$event_id %in% rmats_events_id$event_id,]
  rownames(rmats_events_inc) <- rmats_events_inc$event_id
  rownames(rmats_events_id) <- rmats_events_id$event_id

  #Add columns with INC_LEVEL for each sample
  for (event in 1:nrow(rmats_events_id)){
    samples_inc <- c(eval(parse(text = paste0("c(",rmats_events_id[event,"IncLevel1"] ,")"))),
                     eval(parse(text = paste0("c(",rmats_events_id[event,"IncLevel2"] ,")"))))
    for(sample in 1:length(rmats_order_selected)){
      rmats_events_id[event,rmats_order_selected[sample]] <- samples_inc[sample]}}
  
  #Mark samples used for RMATS in PCA
   rmats_pca_marked<-rmats_pca_marked<-pca_plot(cts_normal_cancer,
                     groupMeta = meta,
                     meta_type = 0,
                     pc1 =1,
                     pc2 = 2,
                     myTitle = 'PCA',
                     markA = rmats_order_selected_1,
                     markB = rmats_order_selected_2
                     ) +
  guides(fill = guide_legend(label.position = "right"), keywidth=1) + 
  theme_bw() + 
  theme(plot.title = element_text(size=40, hjust = 0.5),
        legend.box.background = element_rect(colour = "black"),
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))
   ggsave(paste0(session_dir,'/PCA_',rmats_selected_table_name, '.png'),
          plot = rmats_pca_marked, width = 15, height = 15)
  }
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~COMBINED RMATS EVENTS    V
```{r}
# list_rmats_events_path <- c("N_C_SE.MATS.JCEC.txt",
#                             "N_PT_SE.MATS.JCEC.txt",
#                             "C_PT_SE.MATS.JCEC.txt")
# 
# rmats_events_combined_inc <- NULL
# 
# for( i in 1:length(list_rmats_events_path))
#   rmats_events_combined_inc <- rbind (rmats_events_combined_inc, 
#                                       read.csv(file =list_rmats_events_path[i], row.names = 1))
# 
# rmats_events_combined_inc <- rmats_events_combined_inc[!duplicated(rmats_events_combined_inc$ID), ]

for_complexHM <- as.matrix(rmats_events_combined_inc[,meta$Sample.Name])
for_complexHM <- for_complexHM[apply(for_complexHM, 1, var, na.rm=TRUE) != 0,] # remove genes with no variance
```
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RMATS HEATMAP    V
```{r}
cell_h <- 3
cell_w <- 1
plot_name <- 'SKIPPED EXON EVENTS'

png(paste0('plots/',plot_name,'.png'),
  height = cell_h*(nrow(for_complexHM) + ncol(Ann_all_t))+150,
  units = "mm",
  res = 300)
complex_hm_breaks <- quantile_breaks(scale_rows(for_complexHM),100)
cHM <- Heatmap(t(scale(t(for_complexHM))),
        column_title = 'SKIPPED EXON EVENTS, |\u0394 PSI| > 0.2, FDR = 0',
        name = 'Expression level',
        col = colorRamp2(complex_hm_breaks, colorRampPalette(c("skyblue","#FFFFBF", "#D73027"))(length(complex_hm_breaks))),
        na_col = 'black',
        cluster_rows = TRUE,
        cluster_columns = TRUE,
        clustering_method_rows = "ward.D",
        clustering_method_columns = "ward.D",
        width = ncol(for_complexHM) * unit(cell_w,'mm'),
        height = nrow(for_complexHM) * unit(cell_h,'mm'),
        row_names_gp = gpar(fontsize = 8),
        column_names_gp = gpar(fontsize = 2),#5
        row_names_max_width = max_text_width(rownames(for_complexHM)),
        heatmap_legend_param = list(direction = 'horizontal', legend_width  = unit(2,'inches')),
        top_annotation = HeatmapAnnotation( df = Ann_all_t[,ann_all_col],
                                            col = annotation_color_list_t,
                                            na_col = 'white',
                                            annotation_name_gp= gpar(fontsize = 12),
                                            simple_anno_size = unit(4, "mm"),
                                            show_legend = c(rep(TRUE,6),rep(FALSE,29),TRUE,TRUE,TRUE),
                                            gp = gpar(col = "grey90"),
                                            annotation_legend_param = 
                                            list('State' = list(at = c('Normal', 'Cancer', 'Thrombus')),
                                                 'Type B intercalated cell' = list(title = 'Cell inclusion',
                                                                                   labels = c("Min", "Half", "Max"))))
        
    )
draw (cHM,heatmap_legend_side = "bottom",annotation_legend_side = "left", legend_grouping = "original")
dev.off()
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~DECONVULVE   V
```{r}
##Download the Mature_Full_v3.h5ad fle from Kidneycellatlas, use python note book on it.
library(Seurat)
library(BisqueRNA)
sc.meta <- read.csv ('data/deconvolution_Mature_full/metadata.csv', row.names = 1)
sc <- Read10X(data.dir = 'data/deconvolution_Mature_full/', strip.suffix = F)

sobj <- CreateSeuratObject(counts = sc, meta.data = sc.meta, names.field = 2, names.delim = '-')

bulk.eset <- Biobase::ExpressionSet(assayData = as.matrix(kirc_cts))
sc.eset <- SeuratToExpressionSet(sobj, delimiter='-',position=1, version="v3")

bis.res <- BisqueRNA::ReferenceBasedDecomposition(bulk.eset, 
                                                  sc.eset,
                                                  markers=NULL, 
                                                  use.overlap=FALSE)
ref.based.estimates <- bis.res$bulk.props
if(!file.exists("Table deconvolution.csv"))
   write.csv(ref.based.estimates, file = "Table deconvolution.csv")
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~CELL TYPE FEATURE    V
```{r}
cnv_hm <- ref.based.estimates
celltypedir <- 'CELL_PROPORTIONS/'

dir.create(celltypedir)
 for (cell in rownames(ref.based.estimates))
 {
   celltype_feature_plot <- feature_plot3(cts_normal_cancer,
                                         inclusion_level = as.numeric(cnv_hm[cell,]),
                                         min_point_size = 0.5,
                                         myTitle = cell,
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Cell Proportions") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1)) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=30, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
  
  ggsave(paste0(celltypedir,cell,'.png'), 
         plot = celltype_feature_plot, width = 15, height = 15)
 }
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~TCGA TYPE NORMALIZATION    V
```{r}
clinical_kirc <- GDCquery_clinic('TCGA-KIRC')
clinical_kirc$deceased <- ifelse(clinical_kirc$vital_status == "Alive", FALSE, TRUE)
clinical_kirc$overall_survival <- ifelse(clinical_kirc$vital_status == "Alive",
                                         clinical_kirc$days_to_last_follow_up,
                                         clinical_kirc$days_to_death)

kirc_query <- GDCquery(project= 'TCGA-KIRC',
                       data.category = 'Transcriptome Profiling',
                       workflow.type = 'STAR - Counts',
                        data.type = 'Gene Expression Quantification',
                       sample.type = c("Solid Tissue Normal","Primary Tumor"))

GDCdownload(kirc_query)
kirc_prepared <- GDCprepare(kirc_query, summarizedExperiment = TRUE)
kirc_prepared<- na.omit(kirc_prepared)

#translate gene id to gene name
id2gene <- data.frame (gene_id = kirc_prepared@rowRanges@elementMetadata@listData[["gene_id"]],
                       gene_name = kirc_prepared@rowRanges@elementMetadata@listData[["gene_name"]])

kirc_matrix <- na.omit(assay(kirc_prepared, 'unstranded'))
kirc_gene_meta <- as.data.frame(rowData(kirc_prepared))
kirc_coldata <- as.data.frame(colData(kirc_prepared))
kirc_coldata$sample_type <- gsub(' ','.',kirc_coldata$sample_type)

kirc_dds_type <- DESeqDataSetFromMatrix(countData = kirc_matrix,
                                   colData = kirc_coldata,
                                   design = ~sample_type)
kirc_dds_type <- DESeq(kirc_dds_type)
kirc_cts_type <- as.data.frame(counts(kirc_dds_type, normalized = TRUE))
kirc_cts_type[,'gene_id'] <- rownames(kirc_cts_type)
kirc_cts_type <- left_join(kirc_cts_type, id2gene, by = 'gene_id')

kirc_vsd <- vst(kirc_dds_type,blind = FALSE)
kirc_matrix_vst <- assay(kirc_vsd)[,kirc_coldata$shortLetterCode == 'TP']
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RBP KAPLAN-MEIER   V
```{r}
kirc_1gene_all <- kirc_matrix_vst %>%
  as.data.frame() %>%
  rownames_to_column(var = 'gene_id') %>%
  gather(key = 'case_id',value = 'counts', -gene_id) %>%
  left_join(., id2gene, by = 'gene_id')

rbp_in_kirc <- gene_lists$`Motif genes`

rbp_in_kirc <- rbp_in_kirc[rbp_in_kirc %in% kirc_1gene_all$gene_name]

km_risk_path <- paste0('plots/KM_RISK_',Sys.Date())

dir.create(km_risk_path)

for (kirc_rbp in rbp_in_kirc)
  {
    kirc_1gene <- kirc_1gene_all %>% filter (gene_name == kirc_rbp)
      
    median_kirc_1gene <- median(kirc_1gene$counts)
    kirc_1gene$strata <- ifelse(kirc_1gene$counts >= median_kirc_1gene, 'HIGH','LOW')
    kirc_1gene$case_id <- gsub('-01.*', '', kirc_1gene$case_id)
    kirc_1gene <- merge(kirc_1gene, clinical_kirc, by.x = 'case_id', by.y = 'submitter_id')
    
    kirc_1fit <- survfit(Surv(overall_survival, deceased) ~ strata, data = kirc_1gene)
    
    gene_km_plot <- ggsurvplot(kirc_1fit,
               palette = c("#D73027", "skyblue"),
               data = kirc_1gene,
               size = 4,
               pval = T,
               pval.size = 15,
               title = kirc_rbp,
               legend = c(.8,.9),
               font.main = 50,
               font.legend = 30,
               font.ytickslab = 30,
               font.xtickslab = 30,
               ggtheme = theme_pubr() + theme(legend.key.size = unit(5,'lines')),
               risk.table = 'absolute',
               fontsize = 15)
    
    gene_km_plot$table$theme$text$size <- 15
    gene_km_plot$table$theme$plot.title$size <- 30
    gene_km_plot$table$theme$axis.text$size <- 30
    gene_km_plot$table$theme$axis.title.x$size <- 30
    
    png(paste0(km_risk_path,'/',kirc_rbp,'.png'),
        units = "mm",
        res = 300)
    
    print(gene_km_plot)
    dev.off()
    print(kirc_rbp)
}
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RMAPS COMBINED WITH FEATURE AND KM   V
```{r}
rmaps_plots_path <- 'plots/rmaps/RMAPS_COMBINED_2024-02-28/'

rbp_plots_path <- 'plots/rbp_feature_hist/'

km_plots_path <- 'plots/KM/KM_RISK_2024-04-16/'

save_path_rmaps_feature <- paste0("plots/FI_RMAPS_FEATURE_KM_HIST",Sys.Date(),'/')

dir.create(save_path_rmaps_feature)

add_rmaps<- list.files(rmaps_plots_path, pattern = "*.png")
add_rbp <- list.files(rbp_plots_path, pattern = "*.png")
add_km <- list.files(km_plots_path, pattern = "*.png")
  
rmaps_names <- unlist(lapply(str_extract_all(add_rmaps,"[^-]+(?=-)|(?<=-)[^-]+"), 
                                 function(name) name[1]))

rbp_names <-   unlist(str_extract_all(add_rbp,".+(?=\\.)"))
#unlist(str_extract_all(add_rbp,"(?<=_)(.*?)(?=\\.)"))

km_names <- rbp_names
                                 
names(add_rmaps) <- rmaps_names
names(add_rbp) <- rbp_names
names(add_km) <- km_names

#Exceptions
names(add_rmaps)[which(rmaps_names == '9G8')] <- 'SRSF7'
names(add_rmaps)[which(rmaps_names == 'BRUNOL4')] <- 'CELF4'
names(add_rmaps)[which(rmaps_names == 'BRUNOL5')] <- 'CELF5'
names(add_rmaps)[which(rmaps_names == 'BRUNOL6')] <- 'CELF6'
names(add_rmaps)[which(rmaps_names == 'HNRPLL')] <- 'HNRNPLL'
names(add_rmaps)[which(rmaps_names == 'HuR')] <- 'ELAVL1'
names(add_rmaps)[which(rmaps_names == 'PTB')] <- 'PTBP1'
names(add_rmaps)[which(rmaps_names == 'SF2')] <- 'SRSF1'
names(add_rmaps)[which(rmaps_names == 'SRp20')] <- 'SRSF3'
names(add_rmaps)[which(rmaps_names == 'SRp40')] <- 'SRSF5'
names(add_rmaps)[which(rmaps_names == 'SRp55')] <- 'SRSF6'
names(add_rmaps)[which(rmaps_names == 'Tra2')] <- 'TRA2B'
add_rmaps['RBFOX2'] <- add_rmaps['RBFOX1']

add_rmaps <-add_rmaps[order(names(add_rmaps))]

# add iso_n notation to isoform of rmaps
iso_table <- table(names(add_rmaps))
for(freq in 1:length(iso_table)){
  if (iso_table[freq] > 1){
    name_index <-  which(names(add_rmaps) == names(iso_table)[freq])[1]
    for(iso in 1:iso_table[freq]){
     names(add_rmaps)[name_index + iso-1] <- paste0(names(iso_table)[freq],'_iso',iso)}}}
 
for (rmaps in names(add_rmaps))
  {
    rmaps_iso0 <- str_extract_all(rmaps, '[^_]+')[[1]][1]
    if(rmaps_iso0 %in% rbp_names){
    to_plot <- c(paste0(rmaps_plots_path,add_rmaps[[rmaps]]),
                 paste0(rbp_plots_path,add_rbp[[rmaps_iso0]]),
                 paste0(km_plots_path,add_km[[rmaps_iso0]]))
    
    plots <- lapply(to_plot,function(plot_me){
    img <- image_read(plot_me)
    rasterGrob(as.raster(img), interpolate = FALSE)})
    
    ggsave(paste0(save_path_rmaps_feature,rmaps,'.png'),
           width=48, height=16,
           marrangeGrob(grobs = plots, ncol=length(to_plot), nrow=1,top=NULL))}
  }
```

PASTE 2 PICTURES TOGETHER   V
```{r}
path1 <- 'plots/Sashimi/GGSASHIMI_SUPP/XPA_chr9_-_97682199_97682430_14.png' #sashimi
path2 <- 'plots/MA_XPA_chr9_-_97682199_124271.png' #feature
path3 <- 'plots/XPA_14.png'

to_plot <- c(path1, path2)
    
plots <- lapply(to_plot,function(plot_me){
  img <- image_read(plot_me)
  #img <- (image_annotate(img,'SE', size = 60,location = "+0+30",color = "black"))
  rasterGrob(as.raster(img), interpolate = FALSE)})
    
ggsave(path3,width=32, height=16,
           marrangeGrob(grobs = plots, ncol=length(to_plot), nrow=1,top=NULL))
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~DESEQ PT-TT    V
```{r fig.width=15, fig.height=15}
cts_pt_tt <- ctstk[,meta$Thrombosis == "Yes"]
meta_pt_tt <- meta[meta$Thrombosis == "Yes", ]
if(!exists("dsq_pt_tt"))
{
  dsq_pt_tt <- DESeqDataSetFromMatrix(countData = cts_pt_tt,
                                      colData = meta_pt_tt,
                                      design= ~ State)
  dsq_pt_tt <- DESeq(dsq_pt_tt)
  ctsnorm_pt_tt <- counts(dsq_pt_tt,normalized=TRUE)
  res_pt_tt <- results(dsq_pt_tt,contrast = c('State','Thrombus','Cancer'))

  #list of genes with p_adj < 0.01 && Log2FC > |1|
  sig_pt_tt <- sigres(res_pt_tt, 0.01, 1)
}

pca_thrombosis<-pcaMethods2(dsq_pt_tt,
           res_pt_tt,
           pc1=1, pc2=2 ,
           method = "all",
           pCutoff = 0.01,
           FCcutoff = 1,
           pSize=4,
           lSize = 2,
           labStyle = 1, # 1no / 2text / 3box / 4box + arrow
           ntop = Inf,
           toScale=T,
           lColor = 0,
           incLevel_size = 1,
           myTitle = 'PCA (all genes)',
           groupMeta = meta_pt_tt)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PAIRS SIMULATION   V
```{r}
 pca_PT_TT <- data.frame(pc1 = pca_thrombosis$data$PC1,
                         pc2 = pca_thrombosis$data$PC2,
                         row.names = pca_thrombosis$data$name)
 pca_PT_TT <- pca_PT_TT[order(rownames(pca_PT_TT)),]

n_pairs <- nrow(pca_PT_TT) / 2

set.seed(1)
pmeans<- NULL
pmeans[1000] <- NA
for(j in 1:999) { 
  c_pair <- sample(1:n_pairs) * 2 - 1 # C sample at odd index
  t_pair <- sample(1:n_pairs) * 2     # T sample at even index
  pmeans[j] <- mean(sapply (1:n_pairs, 
                            function(i) dist2(pca_PT_TT[c_pair[i],],
                                              pca_PT_TT[t_pair[i],])))}

pmeans[1000] <- mean(sapply (1:n_pairs, 
                             function(i) dist2(pca_PT_TT[i*2-1,],
                                               pca_PT_TT[i*2,])))
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~DENSITY PT-TT    V
```{r}
myhist <- ggplot(as.data.frame(pmeans), aes(x = pmeans)) +
  geom_histogram(aes(y = after_stat(density),colour = 'Simulated'), 
                 binwidth=01, fill="gray",alpha=.5,linewidth = 0.001) +
  geom_density(alpha=.2, fill = 'skyblue')  + 
  xlab("Mean distance") + 
  ylab("Density") +
    theme_bw() + 
    theme(plot.title = element_text(size=40, hjust = 0.5),
          legend.position = 'bottom',
          legend.direction = 'horizontal',
          legend.title = element_text(size=30),
          legend.text = element_text(size=30, hjust = 0.0),
          legend.key.width = unit(1, "inches"),
          axis.title = element_text(size=30),
          axis.text.x=element_text(size=30),
          axis.text.y=element_text(size=30))

myh2 <- myhist +
  geom_point(aes(x=pmeans[1000],
                 y=ggplot_build(myhist)$data[[1]][1,"density"],
                 colour="True"),size = 5) + 
  scale_color_manual(name = '',
                     breaks=c('Simulated','True'),
                     values=c('Simulated'='skyblue', 'True'='#D73027')) +
  theme_bw() + guides(color=guide_legend(override.aes=list(fill=NA))) + 
   ggtitle("Distance between Primary Tumor\nand Thrombus extension") + 
    theme(plot.title = element_text(size=40, hjust = 0.0),
          legend.box.background = element_rect(colour = "black"),
          legend.position = 'bottom',
          legend.direction = 'horizontal',
          legend.title = element_text(size=30),
          legend.text = element_text(size=30, hjust = 0.0),
          #legend.key.width = unit(1, "inches"),
          axis.title = element_text(size=30),
          axis.text.x=element_text(size=30),
          axis.text.y=element_text(size=30))

ggsave("plots/PCA_PT_TT_HIST1.png", width = 10, height = 10)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RBP HISTOGRAM    V
```{r}
x1 <- log2(cts_normal_cancer + 1)#for_complexHM#
#hist_list <-  unique(gene_lists$`Motif genes`) #rownames(x1) 
bw <- 0.1
for(rbp in hist_list){
  hist_cancer <- data.frame (counts = x1[rbp,patient_state(meta) == 'Cancer'])
  hist_normal <- data.frame (counts = x1[rbp,meta$Normal == 'Yes'])
  hist_thrombus <- data.frame (counts = x1[rbp,meta$Thrombosis == 'Yes'])
  #rbp_hist <-
   ggplot() + #ggtitle(paste0(rbp, ' histogram'))+
     xlab(expression('Log'['2']~'(counts+1)')) + #xlab('Inclusion level') + #
     ylab('Density') + 
      geom_histogram(data  = hist_cancer, 
                     aes(x = counts, y = after_stat(density),fill = 'Cancer'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05)+
      
      geom_histogram(data  = hist_normal, 
                     aes(x = counts, y = after_stat(density),fill = 'Normal'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05) +
     
     geom_histogram(data  = hist_thrombus, 
                     aes(x = counts, y = after_stat(density),fill = 'Thrombus'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05) +
      
      scale_fill_manual(name = '',
                      breaks=c('Normal','Cancer','Thrombus'),
                      values=c('Cancer' =  '#7CAE00',
                               'Normal'='#F8766D',
                               'Thrombus' =  '#00BFC4'
                               )) + theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           #panel.background = element_rect(fill='transparent'),
           plot.background = element_rect(fill='transparent', color=NA),
           legend.position = 'top',
           legend.direction = 'horizontal',
           legend.title = element_text(size=60),
           legend.text = element_text(size=60, hjust = 0.0),
           #legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=60),
           axis.text.x=element_text(size=60),
           axis.text.y=element_text(size=60))
    
 ggsave(paste0('plots/hist5/',rbp,'.png'), width = 15, height = 15,bg='transparent')}
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~RBP WITH HIST    V
```{r}
for (rbp_gene in unique(gene_lists$`Motif genes`))
{
rbp_feature_plot <- feature_plot(cts_normal_cancer,
                                 min_point_size = 0.2,
                                 inclusion_level = as.numeric(cts_normal_cancer[rbp_gene,]),
                                 color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                 myTitle = rbp_gene,
                                 name_t = 'Expression level')  +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=30, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))

hist_file <- paste0('plots/hist5/',rbp_gene,'.png')
my_plot_2 <- ggdraw() + draw_plot(rbp_feature_plot) +
  draw_image(hist_file, x = 1, y = 1, hjust = 1, vjust = 1, halign = 1, valign = 1, width = 0.33, height = 0.33)

ggsave(paste0('plots/rbp_feature_hist5/',rbp_gene,'.png'),
         plot = my_plot_2, width = 15, height = 15)
}
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~PCA RMATS CSTAT 0    V
```{r, fig.width= 300}
pca_simple <- pca_plot(na.omit(rmats_events_combined_inc[,colnames(cts_normal_cancer)]),
                     groupMeta = meta[colnames(cts_normal_cancer),],
                     myTitle = 'PCA'
                     ) 

pca_simple3 <- pca_simple + 
  guides(fill = guide_legend(label.position = "right"), keywidth=1) + 
  theme_bw()  + 
  theme(plot.title = element_text(size=40, hjust = 0.5),
         plot.margin = margin(1, 1, 1, 1, "cm"),
        legend.box.background = element_rect(colour = "black"),
        legend.position = 'bottom',
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))

ggsave("plots/PCA_RMATS_C0.10.png", width = 15, height = 15,limitsize = FALSE)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~CORR PLOT   V
```{r}
genes<- c('BIRC5','BIRC6')
State <- factor(patient_state(meta), levels = c('Normal','Cancer', 'Primary Tumor', 'Thrombus Tumor'))
for(gene in genes[genes %in% rownames(cts_normal_cancer)]){
  gene_cor_plot <- ggplot() + geom_point(aes(x = log2(cts_normal_cancer['TOP2A',] +1),
                            y = log2(cts_normal_cancer[gene,] +1),
                      fill = State), size = 7,
                      shape = 21,color = 'black') +
  theme_bw() + 
  guides(fill = guide_legend(label.position = "right", title=''), keywidth=1) +
    xlab (expression('Log'['2']~'(TOP2A+1)')) + 
    ylab(bquote('Log'['2']~'('~ .(gene) ~'+1)')) + 
    
    theme(plot.title = element_text(size=40, hjust = 0.5),
        legend.box.background = element_rect(colour = "black"),
        legend.position = 'bottom',
        #legend.direction = 'horizontal',
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        #legend.key.width = unit(1, "inches"),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))

ggsave(paste0('plots/cor50_',gene,'.png'), 
         plot = gene_cor_plot, width = 15, height = 15)
}
```

CORR PLOT GENE CELL TYPE    V
```{r}
genes<- c('BIRC5')
State <- factor(patient_state(meta), levels = c('Normal','Cancer', 'Primary Tumor', 'Thrombus Tumor'))
for(gene in genes[genes %in% rownames(cts_normal_cancer)]){
  gene_cor_plot <- ggplot() + 
    geom_point(aes(x = (cts_normal_cancer[gene,colnames(ref.based.estimates)]),
                   y = as.numeric(ref.based.estimates['Proliferating Proximal Tubule',]*100),
                   fill = State), 
               size = 7,
               shape = 21,
               color = 'black') +
  theme_bw() + 
  guides(fill = guide_legend(label.position = "right", title=''), keywidth=1) +
    ylab ('Proliferating Proximal Tubule proportion') + 
    #xlab(bquote('Log'['2']~'('~ .(gene) ~'+1)')) + 
    
    theme(plot.title = element_text(size=40, hjust = 0.5),
        legend.box.background = element_rect(colour = "black"),
        legend.position = 'bottom',
        #legend.direction = 'horizontal',
        legend.title = element_blank(),
        legend.text = element_text(size=30, margin = margin(r = 40, unit = "pt")),
        #legend.key.width = unit(1, "inches"),
        axis.title = element_text(size=30),
        axis.text.x=element_text(size=30),
        axis.text.y=element_text(size=30))

ggsave(paste0('plots/cor_pro_prox1_',gene,'.png'), 
         plot = gene_cor_plot, width = 15, height = 15)
}
```

SASHIMI HIST    V
```{r}
x1 <- rmats_events_combined_inc
rownames(x1) <- x1$ID

bw <- 0.025
for(rbp in hist_list){
  hist_cancer <- data.frame (counts = as.numeric(x1[rbp,meta$Sample.Name[patient_state(meta) == 'Cancer']]))
  hist_normal <- data.frame (counts = as.numeric(x1[rbp,meta$Sample.Name[meta$Normal == 'Yes']]))
  hist_thrombus <- data.frame (counts = as.numeric(x1[rbp,meta$Sample.Name[meta$Thrombosis == 'Yes']]))
  #rbp_hist <-
   ggplot() + #ggtitle(paste0(rbp, ' histogram'))+
     xlab('Proportions') + 
     ylab('Density') + 
      geom_histogram(data  = hist_cancer, 
                     aes(x = counts, y = after_stat(density),fill = 'Cancer'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05)+
      
      geom_histogram(data  = hist_normal, 
                     aes(x = counts, y = after_stat(density),fill = 'Normal'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05) +
     
     geom_histogram(data  = hist_thrombus, 
                     aes(x = counts, y = after_stat(density),fill = 'Thrombus'),
                     colour = 'black',binwidth=bw,alpha=.3,linewidth = 0.05) +
      
      scale_fill_manual(name = '',
                      breaks=c('Normal','Cancer','Thrombus'),
                      values=c('Cancer' =  '#7CAE00',
                               'Normal'='#F8766D',
                               'Thrombus' =  '#00BFC4'
                               )) + theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           #panel.background = element_rect(fill='transparent'),
           plot.background = element_rect(fill='transparent', color=NA),
           legend.position = 'top',
           legend.direction = 'horizontal',
           legend.title = element_text(size=60),
           legend.text = element_text(size=60, hjust = 0.0),
           #legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=60),
           axis.text.x=element_text(size=60),
           axis.text.y=element_text(size=60))
    
 ggsave(paste0('plots/hist-prop/',rbp,'.png'), width = 15, height = 15,bg='transparent')}
```

CELL TYPE PROPORTIONS REVISED   V
```{r}
for (cell in rownames(ref.based.estimates))
 {
   celltype_feature_plot <- feature_plot2(cts_normal_cancer,
                                         inclusion_level = as.numeric(cnv_hm[cell,]),
                                         min_point_size = 0.5,
                                         myTitle = cell,
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Cell Proportions") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1)) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=30, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
  
  ggsave(paste0(celltypedir,cell,'.png'), 
         plot = celltype_feature_plot, width = 15, height = 15)
 }
```

RBP EXPRESSION REVISED    V
```{r}
rbp_new_dir <- 'RBP_NEW/'
for (cell in unique(gene_lists$`Motif genes`))
 {
   rbp_feature_plot <- feature_plot3(cts_normal_cancer,
                                         inclusion_level = as.numeric(log2(cts_normal_cancer[cell,]+1)),
                                         min_point_size = 0.2,
                                         myTitle = cell,
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Expression level") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1,
                                   title = expression('Log'['2']~'(counts+1)'))) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=30, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
   
   hist_file <- paste0('plots/hist5/',cell,'.png')
   my_plot_2 <- ggdraw() + 
     draw_plot(rbp_feature_plot) +
     draw_image(hist_file, x = 1, y = 1, hjust = 1, vjust = 1, halign = 1, valign = 1, width = 0.33, height = 0.33)

ggsave(paste0('plots/rbp_feature_hist7/',cell,'.png'),
         plot = my_plot_2, width = 15, height = 15)
 }
```

RMATS EXPRESSION REVISED
```{r}
#rmats_events_combined_inc1 <- rmats_events_combined_inc
#rownames(rmats_events_combined_inc1) <- 1:nrow(rmats_events_combined_inc1)
#rmats_new_dir <- 'RMATS_NEW/'

#r2 <- read.csv('data/rmats/mini_rmats_se.csv')
#rownames(r2) <- 1:nrow(r2)

#rmats_events_combined_inc1 <- r2
cells <- #c(172, 180, 181, 184)#c(140,167,96,179) hagh, upp1, ccdc50, klhl5
for (cell in cells)
 {
   rmats_feature_plot <- feature_plot3(cts_normal_cancer,
                                         inclusion_level = as.numeric(rmats_events_combined_inc1[cell,meta$Sample.Name]),
                                         min_point_size = 0.5,
                                         myTitle = rmats_events_combined_inc1[cell,'geneSymbol'],
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Inclusion level") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1)) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=25, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
  
  #hist_file <- paste0('plots/Volcano_Venn_Hist_PCA_ANOVA/sashimi-hist/',rmats_events_combined_inc[cell,'geneSymbol'],'.png')
  #my_plot_2 <- ggdraw() + draw_plot(rmats_feature_plot) +
  #draw_image(hist_file, x = 1, y = 1, hjust = 1, vjust = 1, halign = 1, valign = 1, width = 0.33, height = 0.33)
   
  ggsave(paste0('plots/rmats-fe-2/',rmats_events_combined_inc[cell,'geneSymbol'],'.png'), 
         plot = rmats_feature_plot, width = 15, height = 15)
 }
```

RMATS EXPRESSION REVISED2
```{r}
#rownames(rmats_events_combined_inc) <- 1:nrow(rmats_events_combined_inc)
#rmats_new_dir <- 'RMATS_NEW/'

#177413 APOPT1
#cells <- c(140,167,96,179)
#c(346,491,545)
event <- 19653
samples_inc <- c(eval(parse(text = paste0("c(",asse[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",asse[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)


cells <- 'UPP1'
for (cell in cells)
 {
   rmats_feature_plot <- feature_plot3(cts_normal_cancer,
                                         inclusion_level = samples_inc[meta$Sample.Name],
                                         min_point_size = 0.5,
                                         myTitle = cells,#rmats_events_combined_inc[cell,'geneSymbol'],
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Inclusion level") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1)) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=25, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
   # + geom_text(label=colnames(cts_normal_cancer),hjust=0, vjust=0, size = 2)
  
  ggsave(paste0('plots/test-',cell,'x.png'), #ggsave(paste0(rmats_new_dir,cell,'.png'), 
         plot = rmats_feature_plot, width = 15, height = 15)
 }
```

RMATS WITH HIST   V
```{r}
cell <- 'FGFR2'
cell2 <- 'FGFR2'
  rmats_feature_plot <- feature_plot3(cts_normal_cancer,
                                         inclusion_level = samples_inc[meta$Sample.Name],#as.numeric(rmats_events_combined_inc[cell2,meta$Sample.Name]),
                                      #samples_inc[meta$Sample.Name],#
                                         min_point_size = 0.5,
                                         myTitle = cell,#rmats_events_combined_inc[cell,'geneSymbol'],
                                         color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                         name_t = "Inclusion level") +
     guides(fill = guide_colourbar(ticks.colour = "black",
                                   ticks.linewidth = 1)) +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=25, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))
  
  hist_file <- paste0('plots/Volcano_Venn_Hist_PCA_ANOVA/sashimi-hist/',cell,'.png')
  my_plot_2 <- ggdraw() + draw_plot(rmats_feature_plot) +
  draw_image(hist_file, x = 1, y = 1, hjust = 1, vjust = 1, halign = 1, valign = 1, width = 0.33, height = 0.33)

ggsave(paste0('plots/test-',cell,'.png'),
       plot = my_plot_2, width = 15, height = 15)
```

RMATS EVENTS FOR PLOT   V
```{r}
################AS3 CALD1
event <- 4371
samples_inc <- c(eval(parse(text = paste0("c(",as3[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",as3[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)
################SE ACTN1
event <- 45293
samples_inc <- c(eval(parse(text = paste0("c(",asse[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",asse[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)
################SE CCDC50
event <- 220303
samples_inc <- c(eval(parse(text = paste0("c(",asse[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",asse[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)
################MXE FGFR2
event <- 43735
samples_inc <- c(eval(parse(text = paste0("c(",as_mxe[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",as_mxe[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)

############### NEXT
event <- 220303
samples_inc <- c(eval(parse(text = paste0("c(",r2[event,"IncLevel1"] ,")"))),
                       eval(parse(text = paste0("c(",r2[event,"IncLevel2"] ,")"))))
names(samples_inc) <- rmats_order_all
samples_inc[is.na(samples_inc)] <- mean(samples_inc, na.rm = T)
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~GENE FEATURE PLOT REVISED
```{r}
genes<- c('NAIP', 'BIRC2', 'BIRC3', 'BIRC5', 'BIRC6')
for(gene in genes[genes %in% rownames(cts_normal_cancer)]){
  n <- cts_normal_cancer[gene,meta$Normal == 'Yes']
  c <- cts_normal_cancer[gene,meta$Normal == 'No' & meta$Thrombosis == 'No']
  th <- cts_normal_cancer[gene,meta$Thrombosis == 'Yes']
gene_feature_plot <- feature_plot3(cts_normal_cancer,
                                  inclusion_level = as.numeric(log2(cts_normal_cancer[gene,] +1)),
                                  color_wheel = c("skyblue","#FFFFBF", "#D73027"),
                                  name_t = expression('Log'['2']~'(counts+1)'),#'Expression level',
                                  myTitle = #paste0('Gene expression ',gene,': \nN = ',
                                            #       format(round(median(n), 2)),'\u00B1', format(round(sd(n), 2)),
                                            #       ' \nC = ', format(round(median(c), 2)), '\u00B1', format(round(sd(c), 2)),
                                            #       ' \nTh = ',format(round(median(th), 2)), '\u00B1', format(round(sd(th), 2)))
                                    gene
                                  )  +
     theme_bw() + 
     theme(plot.title = element_text(size=40, hjust = 0.5),
           legend.position = 'bottom',
           legend.direction = 'horizontal',
           legend.title = element_text(size=30),
           legend.text = element_text(size=20, hjust = 0.0),
           legend.key.width = unit(1, "inches"),
           axis.title = element_text(size=30),
           axis.text.x=element_text(size=30),
           axis.text.y=element_text(size=30))

ggsave(paste0('plots/',gene,'.png'), 
         plot = gene_feature_plot, width = 15, height = 15)}
```