# ccRCC_Wang_2020
Bioinformatics research . Code used with Wang's data on clear-cell renal cell carcinoma.  
Wang's article can be found at https://doi.org/10.1038/s41467-020-14601-9

### MATERIALS AND METHODS
Datasets and Preprocessing:
RNA sequencing reads generated by Wang et al. were downloaded from the NCBI Sequence Read Archive (SRA) (accession code PRJNA596338) using SRA tools. Patients' clinical metadata was obtained from the journal's website. [STAR](https://doi.org/10.1093/bioinformatics/bts635) (version 2.7.10b) was used to align the reads to the human reference genome (hg38) and to obtain a raw counts matrix. [DESeq2](https://doi.org/10.1186/s13059-014-0550-8) (version 1.42.0) was used to obtain a normalized counts matrix. 

### Cell Deconvolution:
[BisqueRNA](https://doi.org/10.1038/s41467-020-15816-6) (version 1.0.5) was used to estimate cell type proportions within each bulk RNA sample. As a reference single-cell RNAseq gene expression dataset, we used the adult human kidney dataset downloaded from the [Kidney Cell Atlas](https://www.kidneycellatlas.org/). Since BisqueRNA requires the single-cell reference as a Seurat object, we used the ‘scanpy’ python library (version 1.9.6) to convert the single-cell reference dataset from H5ad format to Market Exchange Format (MEX) that can be read by [Seurat](https://doi.org/10.1038/s41587-023-01767-y) (version 5.0.1). Heatmaps were generated using the ‘complexHeatmap’ R package (version 2.18.0).

### Alternative splicing:
[rMATS-turbo](https://doi.org/10.1038/s41596-023-00944-2) (version 4.12) was used to detect alternatively spliced transcripts. In each rMATS run we compared two sets of samples chosen according to their position in PCA latent space. Additionally, to obtain the inclusion levels across all samples and all splicing events, we performed a separate rMATS run with the option ‘--cstat 0’ on all samples. Alternative splicing in selected transcripts was visually validated using the [Integrated Genome Viewer (IGV)](https.//doi.org/10.1093/bioinformatics/btac830). Sashimi plots were generated using the [‘ggsashimi’ command-line tool](https://doi.org/10.1371/journal.pcbi.1006360) (version 1.1.5) with the options ‘--shrink’ and ‘-A median_j’.

### Motif enrichment analysis:
The outputs of rMATS for each comparison were used as input to [rMAPS](https://doi.org/10.1093/nar/gkaa237) (version 2.0.1) in order to identify putative splicing regulators. rMAPS searches for enrichment in binding motifs associated with known RNA binding proteins (RBPs) within the sequences of alternatively spliced transcripts in the vicinity of alternatively spliced exons.

### Survival analysis:
Clear cell RCC RNAseq expression data (TCGA-KIRC, 541 samples) and associated metadata were downloaded from the [TCGA database](https://doi.org/10.1038/ng.2764) using the [‘TCGAbiolinks’ R package](https://doi.org/10.1093/nar/gkv1507) (version 2.30.0). Kaplan-Meier curves for RBPs, as well as p-values for association between their expression levels and patient survival, were plotted using the ‘ggsurvplot’ function from the survminer R package (version 0.4.9).

